packages/dataviews/src/components/dataviews-view-config/index.tsx

```diff
@@ -395,83 +375,77 @@ function FieldItem( {
 	);
 }
 
-function FieldList( {
-	fields,
-	fieldIds,
-	mandatoryFields,
-	viewFields,
-	view,
-	onChangeView,
-}: Omit< Parameters< typeof FieldItem >[ 0 ], 'fieldId' > & {
-	fieldIds: string[];
-} ) {
-	return fieldIds.map( ( fieldId ) => (
-		<FieldItem
-			key={ fieldId }
-			fields={ fields }
-			fieldId={ fieldId }
-			mandatoryFields={ mandatoryFields }
-			viewFields={ viewFields }
-			view={ view }
-			onChangeView={ onChangeView }
-		/>
-	) );
-}
-
 function FieldControl() {
 	const { view, fields, onChangeView } = useContext( DataViewsContext );
-	const mandatoryFields = useMemo(
-		() => getMandatoryFields( view ),
+
+	const visibleFieldIds = useMemo(
+		() => getVisibleFieldIds( view, fields ),
+		[ view, fields ]
+	);
+	const hiddenFieldIds = useMemo(
+		() => getHiddenFieldIds( view, fields ),
+		[ view, fields ]
+	);
+	const notHidableFieldIds = useMemo(
+		() => getNotHidableFieldIds( view ),
 		[ view ]
 	);
-	const viewFields = view.fields || fields.map( ( field ) => field.id );
-	const visibleFields = view.fields;
-	const hiddenFields = useMemo( () => {
-		const nonViewFieldsList = fields
-			.filter(
-				( field ) =>
-					! viewFields.includes( field.id ) &&
-					! mandatoryFields?.includes( field.id )
-			)
-			.map( ( field ) => field.id );
 
-		if ( view.type !== LAYOUT_TABLE ) {
-			return nonViewFieldsList;
-		}
-		const nonViewFieldsAndNonCombinedList = nonViewFieldsList.filter(
-			( fieldId ) => {
-				return ! view.layout?.combinedFields?.some( ( combinedField ) =>
-					combinedField.children.includes( fieldId )
-				);
-			}
-		);
-		const nonViewFieldsCombinedFieldsList =
-			view.layout?.combinedFields
-				?.filter(
-					( combinedField ) =>
-						! viewFields.includes( combinedField.id )
-				)
-				.map( ( combinedField ) => combinedField.id ) || [];
-		return [
-			...nonViewFieldsAndNonCombinedList,
-			...nonViewFieldsCombinedFieldsList,
-		];
-	}, [ view, mandatoryFields, fields, viewFields ] );
+	const visibleFields = fields
+		.filter( ( { id } ) => visibleFieldIds.includes( id ) )
+		.map( ( { id, label, enableHiding } ) => {
+			return {
+				id,
+				label,
+				index: visibleFieldIds.indexOf( id ),
+				isVisible: true,
+				isHidable: notHidableFieldIds.includes( id )
+					? false
+					: enableHiding,
+			};
+		} );
+	if ( view.type === LAYOUT_TABLE && view.layout?.combinedFields ) {
```

Comment:
I think this works for now, but my thinking is that ultimately, we should find a way for the layouts to actually "inject" logic here. Or (which is what I had in mind) to have some kind of a slot and the "properties" area be specific per layout. For instance in "table", we should be able to "combine" fields if we want.

Reply:
I agree. If we aim to make the layouts "registerable/unregisterable" by 3rd parties we cannot have this kind of logic in the view config (or any other component).

I've looked at the current usage of layouts in this component and this is what I see:

- [grid layout check](https://github.com/WordPress/gutenberg/blob/2c2d801894f9c002d5bb467abea8027b09aa061b/packages/dataviews/src/components/dataviews-view-config/index.tsx#L528) for show/hide the density picker.
- [table layout check](https://github.com/WordPress/gutenberg/blob/2c2d801894f9c002d5bb467abea8027b09aa061b/packages/dataviews/src/components/dataviews-view-config/index.tsx#L273) to show/hide the ability to reorder fields.
- [table layout check](https://github.com/WordPress/gutenberg/blob/2c2d801894f9c002d5bb467abea8027b09aa061b/packages/dataviews/src/components/dataviews-view-config/index.tsx#L407) to inject "virtual fields" (aka combined).

All of them could be solved by providing a `layout.supports` thing somewhere. Using that approach, it'd still be DataViews' responsibility to define which "features" are supported.

> Or (which is what I had in mind) to have some kind of a slot and the "properties" area be specific per layout. For instance in "table", we should be able to "combine" fields if we want.

This approach would be a lot more open. However, every layout would have to implement the same actions (reordering, visibility, etc.).

Reply:
Actually, `layout.supports` could also be a nice thing to have for consumers that want to "lock in" a bit more the user experience.

Reply:
> This approach would be a lot more open. However, every layout would have to implement the same actions (reordering, visibility, etc.).

Yes, but for list view we have badges..., for grid we have columns and rows...